packages/prover/src/proof_provider/payload_store.ts

```diff
@@ -0,0 +1,154 @@
+import {Api} from "@lodestar/api";
+import {allForks, capella} from "@lodestar/types";
+import {MAX_PAYLOAD_HISTORY} from "../constants.js";
+import {getExecutionPayloadForBlockNumber, getExecutionPayloads} from "../utils/consensus.js";
+import {bufferToHex, hexToNumber} from "../utils/conversion.js";
+import {OrderedMap} from "./ordered_map.js";
+
+type BlockELRoot = string;
+type BlockCLRoot = string;
+
+/**
+ * The in-memory store for the execution payloads to be used to verify the proofs
+ */
+export class PayloadStore {
+  // We store the block numbers only for finalized blocks
+  private finalizedRoots = new OrderedMap<BlockELRoot>();
+
+  // Unfinalized blocks are stored by the roots of the beacon chain
+  private unfinalizedRoots = new Map<BlockCLRoot, BlockELRoot>();
+
+  // Payloads store with BlockELRoot as key
+  private payloads = new Map<BlockELRoot, allForks.ExecutionPayload>();
+
+  private latestBlockRoot: BlockELRoot | null = null;
+
+  constructor(private opts: {api: Api}) {}
+
+  get finalized(): allForks.ExecutionPayload | undefined {
+    const finalizedMaxRoot = this.finalizedRoots.get(this.finalizedRoots.max);
+    if (finalizedMaxRoot) {
+      return this.payloads.get(finalizedMaxRoot);
+    }
+
+    return undefined;
+  }
+
+  get latest(): allForks.ExecutionPayload | undefined {
+    if (this.latestBlockRoot) {
+      return this.payloads.get(this.latestBlockRoot);
+    }
+
+    return undefined;
+  }
+
+  async get(blockId: number | string): Promise<allForks.ExecutionPayload | undefined> {
+    // Given block id is a block hash in hex (32 bytes root takes 64 hex chars + 2 for 0x prefix)
+    if (typeof blockId === "string" && blockId.startsWith("0x") && blockId.length === 64 + 2) {
+      return this.payloads.get(blockId);
+    }
+
+    // Given block id is a block number in hex
+    if (typeof blockId === "string" && blockId.startsWith("0x")) {
+      return this.getOrFetchFinalizedPayload(hexToNumber(blockId));
+    }
+
+    // Given block id is a block number in decimal string
+    if (typeof blockId === "string" && !blockId.startsWith("0x")) {
+      return this.getOrFetchFinalizedPayload(parseInt(blockId, 10));
+    }
+
+    // Given block id is a block number in decimal
+    if (typeof blockId === "number") {
+      return this.getOrFetchFinalizedPayload(blockId);
+    }
+
+    return undefined;
+  }
+
+  async getOrFetchFinalizedPayload(blockNumber: number): Promise<allForks.ExecutionPayload | undefined> {
+    if (blockNumber > this.finalizedRoots.max) {
+      throw new Error(
+        `Block number ${blockNumber} is higher than the latest finalized block number. We recommend to use block hash for unfinalized blocks.`
+      );
+    }
+
+    let blockELRoot = this.finalizedRoots.get(blockNumber);
```

Comment:
a comment here would be nice:
`\\ check if we have payload cached locally else fetch from api`